api/datastore/mysql/mysql.go

```diff
@@ -0,0 +1,615 @@
+package mysql
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	"github.com/Sirupsen/logrus"
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/iron-io/functions/api/datastore/internal/datastoreutil"
+	"github.com/iron-io/functions/api/models"
+	_ "github.com/lib/pq"
+)
+
+const routesTableCreate = `CREATE TABLE IF NOT EXISTS routes (
+	app_name varchar(256) NOT NULL,
+	path varchar(256) NOT NULL,
+	image varchar(256) NOT NULL,
+	format varchar(16) NOT NULL,
+	maxc int NOT NULL,
+	memory int NOT NULL,
+	timeout int NOT NULL,
+	type varchar(16) NOT NULL,
+	headers text NOT NULL,
+	config text NOT NULL,
+	PRIMARY KEY (app_name, path)
+);`
+
+const appsTableCreate = `CREATE TABLE IF NOT EXISTS apps (
+    name varchar(256) NOT NULL PRIMARY KEY,
+	config text NOT NULL
+);`
+
+const extrasTableCreate = `CREATE TABLE IF NOT EXISTS extras (
+    id varchar(256) NOT NULL PRIMARY KEY,
+	value varchar(256) NOT NULL
+);`
+
+const routeSelector = `SELECT app_name, path, image, format, maxc, memory, type, timeout, headers, config FROM routes`
+
+type rowScanner interface {
+	Scan(dest ...interface{}) error
+}
+
+type rowQuerier interface {
+	QueryRow(query string, args ...interface{}) *sql.Row
+}
+
+/*
+MySQLDatastore ...
+*/
+type MySQLDatastore struct {
+	db *sql.DB
+}
+
+/*
+New ...
+*/
+func New(url *url.URL) (models.Datastore, error) {
+	u := fmt.Sprintf("%s@%s%s", url.User.String(), url.Host, url.Path)
+	db, err := sql.Open("mysql", u)
+	if err != nil {
+		return nil, err
+	}
+
+	err = db.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	maxIdleConns := 30
+	db.SetMaxIdleConns(maxIdleConns)
+	logrus.WithFields(logrus.Fields{"max_idle_connections": maxIdleConns}).Info("MySQL dialed")
+
+	pg := &MySQLDatastore{
+		db: db,
+	}
+
+	for _, v := range []string{routesTableCreate, appsTableCreate, extrasTableCreate} {
+		_, err = db.Exec(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return datastoreutil.NewValidator(pg), nil
+}
+
+/*
+InsertApp ...
+*/
+func (ds *MySQLDatastore) InsertApp(ctx context.Context, app *models.App) (*models.App, error) {
+	var cbyte []byte
+	var err error
+
+	if app == nil {
+		return nil, models.ErrDatastoreEmptyApp
+	}
+
+	if app.Name == "" {
+		return nil, models.ErrDatastoreEmptyAppName
+	}
+
+	if app.Config != nil {
+		cbyte, err = json.Marshal(app.Config)
+		if err != nil {
+			return nil, err
+		}
+	}
+	stmt, err := ds.db.Prepare("INSERT apps SET name=?,config=?")
+
+	_, err = stmt.Exec(app.Name, string(cbyte))
+
+	if err != nil {
+		return nil, models.ErrAppsAlreadyExists
```

Comment:
Not necessarily?
Is there a mysql specific error that can be checked for like with postgres?